src/viewport_renderer.rs

```diff
@@ -0,0 +1,402 @@
+                    rpass.set_index_buffer(&index_buffer, 0);
+                    rpass.draw_indexed(0..*index_count, 0, 0..1);
+                } else {
+                    rpass.draw(0..*vertex_count, 0..1);
+                }
+            } else {
+                log::warn!("Geometry with id {} does not exist in this renderer.", id.0);
+            }
+        }
+    }
+
+    fn create_matrix(aspect_ratio: f32, view_matrix: &Matrix4<f32>) -> Matrix4<f32> {
+        const FOVY: f32 = std::f32::consts::FRAC_PI_3;
+        const ZNEAR: f32 = 0.01;
+        const ZFAR: f32 = 1000.0;
+
+        // Vulkan (and therefore wgpu) has different NDC and
+        // clip-space semantics than OpenGL: Vulkan is right-handed, Y
+        // grows downwards. The easiest way to keep everything working
+        // as before and use all the libraries that assume OpenGL is
+        // to apply a correction to the projection matrix which
+        // normally changes the right-handed OpenGL world-space to
+        // left-handed OpenGL clip-space (apart from actually
+        // performing the projection).
+        // https://matthewwellings.com/blog/the-new-vulkan-coordinate-system/
+        #[rustfmt::skip]
+        let wgpu_correction_matrix = Matrix4::new(
+            1.0,  0.0,  0.0,  0.0,
+            0.0, -1.0,  0.0,  0.0,
+            0.0,  0.0,  0.5,  0.0,
+            0.0,  0.0,  0.5,  1.0,
+        );
+        let projection_matrix = Matrix4::new_perspective(aspect_ratio, FOVY, ZNEAR, ZFAR);
+
+        wgpu_correction_matrix * projection_matrix * view_matrix
+    }
+
+    fn upload_uniform_matrix_buffer(
+        device: &mut wgpu::Device,
+        uniform_buffer: &wgpu::Buffer,
+        matrix: &Matrix4<f32>,
+    ) {
+        let uniform_count = 16;
+        let uniform_buffer_size = mem::size_of::<Matrix4<f32>>() as u64;
+
+        let tmp_buffer = device
+            .create_buffer_mapped(uniform_count, wgpu::BufferUsage::TRANSFER_SRC)
+            .fill_from_slice(matrix.as_slice());
+
+        let mut encoder =
+            device.create_command_encoder(&wgpu::CommandEncoderDescriptor { todo: 0 });
+        encoder.copy_buffer_to_buffer(&tmp_buffer, 0, uniform_buffer, 0, uniform_buffer_size);
+        device.get_queue().submit(&[encoder.finish()]);
+    }
+
+    fn create_depth_texture(device: &wgpu::Device, width: u32, height: u32) -> wgpu::Texture {
+        device.create_texture(&wgpu::TextureDescriptor {
+            size: wgpu::Extent3d {
+                width,
+                height,
+                depth: 1,
+            },
+            array_layer_count: 1,
+            mip_level_count: 1,
+            sample_count: 1,
+            dimension: wgpu::TextureDimension::D2,
+            format: DEPTH_FORMAT,
+            usage: wgpu::TextureUsage::OUTPUT_ATTACHMENT,
+        })
+    }
+}
+
+struct GeometryDescriptor {
+    vertices: (wgpu::Buffer, u32),
+    indices: Option<(wgpu::Buffer, u32)>,
+}
```

Comment:
It'd be probably better if this was on top of this file amongst other structs.

Reply:
It's private and always will be, though. I can do it if you insist, but I usually try to put most public things towards the top of the file (except maybe consts)